packages/gatsby/src/utils/node-manifest.ts

```diff
@@ -0,0 +1,302 @@
+import { ErrorId } from "./../../../gatsby-cli/src/structured-errors/error-map"
+import { getNode } from "./../redux/nodes"
+import { IGatsbyPage, INodeManifest } from "./../redux/types"
+import reporter from "gatsby-cli/lib/reporter"
+import { store } from "../redux/"
+import { internalActions } from "../redux/actions"
+import path from "path"
+import fs from "fs-extra"
+
+interface INodeManifestPage {
+  path?: string
+}
+
+/**
+ * This it the output after processing calls to the public unstable_createNodeManifest action
+ */
+interface INodeManifestOut {
+  page: INodeManifestPage
+  node: {
+    id: string
+  }
+  foundPageBy: FoundPageBy
+}
+
+type FoundPageBy =
+  | `ownerNodeId`
+  | `context.id`
+  | `queryTracking`
+  | `filesystem-route-api`
+  | `none`
+
+/**
+ * Finds a final built page by nodeId
+ *
+ * Note that this function wont work properly in `gatsby develop`
+ * since develop no longer runs all page queries when creating pages.
+ * We use the node id to query mapping to find the right page but
+ * this mapping only exists once you've visited a page in your browser.
+ * When this fn is being used for routing to previews the user wont necessarily have
+ * visited the page in the browser yet.
+ */
+async function findPageOwnedByNodeId({
+  nodeId,
+}: {
+  nodeId: string
+}): Promise<{
+  page: INodeManifestPage
+  foundPageBy: FoundPageBy
+}> {
+  const state = store.getState()
+  const { pages, nodes } = state
+  const { byNode } = state.queries
+
+  // in development queries are run on demand so we wont have an accurate nodeId->pages map until those pages are visited in the browser. We want this mapping before the page is visited in the browser so we can route to the right page in the browser.
+  // So in development we can just use the Map of all pages (pagePath -> pageNode)
+  // but for builds (preview inc builds or regular builds) we will have a full map
+  // of all nodeId's to pages they're queried on and we can use that instead since it
+  // will be a much smaller list of pages, resulting in better performance for large sites
+  const usingPagesMap: boolean =
+    !!process.env._GATSBY_INTERNAL_TEST_NODE_MANIFEST_AS_DEVELOP ||
+    `development` === process.env.NODE_ENV
+
+  const pagePathSetOrMap = usingPagesMap
+    ? // this is a Map of page path to page node
+      pages
+    : // this is a Set of page paths
+      byNode?.get(nodeId)
+
+  // the default page path is the first page found in
+  // node id to page query tracking
+  let pagePath = byNode?.get(nodeId)?.values()?.next()?.value
+
+  let foundPageBy: FoundPageBy = pagePath ? `queryTracking` : `none`
+
+  // but if we have more than one page where this node shows up
+  // we need to try to be more specific
+  if (pagePathSetOrMap && pagePathSetOrMap.size > 1) {
+    let ownerPagePath: string | undefined
+    let foundOwnerNodeId = false
+
+    // for each page this nodeId is queried in
+    pagePathSetOrMap.forEach((...args) => {
```

Comment:
Small nitpick

you could try the following to make it more readable. Both `.values()` operator return a value iterator that you can loop through. Doesn't matter if it's a Set or a Map.
```suggestion
    for (const value of pagePathSetOrMap.values())
```